_posts/2020-10-23-deconvolution.md

```diff
@@ -0,0 +1,446 @@
+(FFT)](https://en.wikipedia.org/wiki/Fast_Fourier_transform). Once this is
+applied we are able to convolve data using a simple multiplication.
+
+As it turns out this FFT transformation is extremely fast on both CPUs and
+GPUs. Similarly the algorithm we can write with FFTs are accelerated.  Despite
+the added step of doing FFTs, the cost of transformation + the cost of the
+algorithm is still lower than performing the original algorithm in real space.
+We (and others before us) found this was the case for Richardson Lucy (on both
+CPUs and GPUs) and performance continued increasing when we parallelized with
+Dask over multiple GPUs.
+
+Help from Open-Source
+---------------------
+
+An FFT RL equivalent has been around for some time and the good folks at the
+[Solar Dynamics Observatory](https://sdo.gsfc.nasa.gov/mission/instruments.php)
+built and shared a NumPy/CuPy implementation as part the [Atmospheric Imaging
+Assembly](https://aiapy.readthedocs.io/en/v0.2.0/_modules/aiapy/psf/deconvolve.html)
+Python package (aiapy).  We slightly modified their implementation to handle 3D
+as well as 2D [Point Spread
+Functions](https://en.wikipedia.org/wiki/Point_spread_function) and to take
+advantage of
+[NEP-18](https://numpy.org/neps/nep-0018-array-function-protocol.html) for
+convenient dispatching of NumPy and CuPy to NumPy and CuPy functions:
+
+
+```python
+def deconvolve(img, psf=None, iterations=20):
+    """
+    """
+    # 2D vs 3D psfs
+    if len(psf.shape) == 2:
+        psf = psf
+    else:
+        psf = psf[0]
+
+    # pad psf with zeros to match image shape
+    # use NumPy for padding
+    pad_l, pad_r = np.divmod(np.array(img[0].shape) - np.array(psf.shape), 2)
+    pad_r += pad_l
+
+    psf = np.pad(psf, tuple(zip(pad_l, pad_r)), 'constant', constant_values=0)
+
+    # padding centers the PSF so we no longer need to roll
+    # Center PSF at pixel (0,0)
+    for i in range(psf.ndim):
+        psf = np.roll(psf, psf.shape[i]//2, axis=i)
+
+    # Convolution requires FFT of the PSF
+    psf = np.fft.rfftn(psf)
+    psf = psf[None]
+
+    psf_conj = psf.conj()
+
+    img_decon = np.copy(img)
+    for _ in range(iterations):
+        ratio = (
+            img / np.fft.irfftn(np.fft.rfftn(img_decon,
+                                             axes=tuple(range(1, psf.ndim))) * psf,
+                                             axes=tuple(range(1, psf.ndim)))
+        )
+        img_decon *= np.fft.irfftn(np.fft.rfftn(ratio,
+                                                axes=tuple(range(1, psf.ndim))) * psf_conj,
+                                                axes=tuple(range(1, psf.ndim)))
+    return img_decon
+```
+
+For a 1.3 GB image we measured the following:
+
+- CuPy ~3 seconds for 20 iterations
+- NumPy ~36 seconds for 2 iterations
+
+We see 10x increase in speed for 10 times the number of iterations -- very
+close to our desired 100X speedup!  Let's explore how this implementation
```

Comment:
Style nit: the "10x" and "100X" have different multiplication signs. I suggest using a consistent style (I use the UTF multiplication sign, "×" (HTML: `&#215;`/`&times;`, Unicode: `U+00D7`)).

Reply:
```suggestion
close to our desired 100x speedup!  Let's explore how this implementation
```